### ISSUE TITLE
An Engine cannot be shared by multiple Contexts in which numpy is imported

### ISSUE THREAD

**Issue by kariya-mitsuru** [graalvm/graalpython#177]
When I import numpy in multiple `Context`s which share an `Engine`, an exception will be thrown.

It can reproduce as follows.

### Environment

- GraalVM, GraalPython: CE 20.3.0 (Java8 or Java11)
- OS: CentOS Linux release 8.1.1911 (Core)
  Linux xxxxxx 4.18.0-147.5.1.el8_1.x86_64 #1 SMP Wed Feb 5 02:00:39 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux

### Java Source
```java
import org.graalvm.polyglot.Context;
import org.graalvm.polyglot.Engine;

public class ShareEngine {
    public static void main(String[] args) {
        try (Engine engine = Engine.create()) {
            for (int i = 0; i < 2; ++i) {
                System.out.println("start " + i);

                try (Context ctx = Context.newBuilder().engine(engine).allowAllAccess(true).build()) {
                    ctx.eval("python", "import site, numpy");
                }

                System.out.println("end " + i);
            }
        }
    }
}
```

### Procedure
```sh
# compile
javac ShareEngine.java

# create venv for clean environment
graalpython -m venv venv-20.3.0.r8

# activate venv
. venv-20.3.0.r8/bin/activate

# install numpy to venv
python -m ginstall install numpy

# execute test application above
java -Dpolyglot.python.SysPrefix=venv-20.3.0.r8 \
     -Dpolyglot.python.IsolateFlag=true \
     -Dpolyglot.python.NoUserSiteFlag=true \
     ShareEngine
```

### Output example on my environment (snipped, full version is attached)
```
start 0
end 0
start 1
Traceback (most recent call last):
null
Exception in thread "main" org.graalvm.polyglot.PolyglotException: java.lang.NullPointerException
	at com.oracle.graal.python.builtins.objects.cext.capi.CApiContext.getTraceMallocDomain(CApiContext.java:233)
	at com.oracle.graal.python.builtins.modules.PythonCextBuiltins$PyTruffleTraceMallocTrack.doCachedDomainIdx(PythonCextBuiltins.java:3326)
	at com.oracle.graal.python.builtins.modules.PythonCextBuiltinsFactory$PyTruffleTraceMallocTrackFactory$PyTruffleTraceMallocTrackNodeGen.execute(PythonCextBuiltinsFactory.java:14856)
	at com.oracle.graal.python.nodes.function.builtins.BuiltinCallNode$BuiltinAnyCallNode.execute(BuiltinCallNode.java:64)
	at com.oracle.graal.python.nodes.function.BuiltinFunctionRootNode.execute(BuiltinFunctionRootNode.java:317)
	at   <llvm> PyDataMem_NEW(../../../../tmp/tmp8ohh58rt/numpy-1.16.4/numpy/core/src/multiarray/alloc.c:223:6276)
	at   <llvm> _npy_alloc_cache(../../../../tmp/tmp8ohh58rt/numpy-1.16.4/numpy/core/src/multiarray/alloc.c:72:2030)
	at   <llvm> npy_alloc_cache(../../../../tmp/tmp8ohh58rt/numpy-1.16.4/numpy/core/src/multiarray/alloc.c:106:2823)
	at   <llvm> PyArray_NewFromDescr_int(../../../../tmp/tmp8ohh58rt/numpy-1.16.4/numpy/core/src/multiarray/ctors.c:1072:28941)
	at   <llvm> PyArray_NewFromDescrAndBase(../../../../tmp/tmp8ohh58rt/numpy-1.16.4/numpy/core/src/multiarray/ctors.c:1181:32080)
	at   <llvm> PyArray_NewFromDescr(../../../../tmp/tmp8ohh58rt/numpy-1.16.4/numpy/core/src/multiarray/ctors.c:1166:31660)
	at   <llvm> PyArray_FromAny(../../../../tmp/tmp8ohh58rt/numpy-1.16.4/numpy/core/src/multiarray/ctors.c:1877:55646)
	at   <llvm> get_ufunc_arguments(../../../../tmp/tmp8ohh58rt/numpy-1.16.4/numpy/core/src/umath/ufunc_object.c:1097:32833)
	at   <llvm> PyUFunc_GenericFunction(../../../../tmp/tmp8ohh58rt/numpy-1.16.4/numpy/core/src/umath/ufunc_object.c:3162:103465)
	at   <llvm> ufunc_generic_call(../../../../tmp/tmp8ohh58rt/numpy-1.16.4/numpy/core/src/umath/ufunc_object.c:4688:152401)
	at <python> __call__(Unknown)
	at <python> <module '/home/kariya/test/polyglot/venv-20.3.0.r8/lib/python3.8/site-packages/numpy-1.16.4-py3.8-linux-x86_64.egg/numpy/lib/scimath.py'>(/home/kariya/test/polyglot/venv-20.3.0.r8/lib/python3.8/site-packages/numpy-1.16.4-py3.8-linux-x86_64.egg/numpy/lib/scimath.py:33:979-989)
	at <python> <module '/home/kariya/test/polyglot/venv-20.3.0.r8/lib/python3.8/site-packages/numpy-1.16.4-py3.8-linux-x86_64.egg/numpy/lib/__init__.py'>(/home/kariya/test/polyglot/venv-20.3.0.r8/lib/python3.8/site-packages/numpy-1.16.4-py3.8-linux-x86_64.egg/numpy/lib/__init__.py:19:422-451)
	at <python> <module '/home/kariya/test/polyglot/venv-20.3.0.r8/lib/python3.8/site-packages/numpy-1.16.4-py3.8-linux-x86_64.egg/numpy/__init__.py'>(/home/kariya/test/polyglot/venv-20.3.0.r8/lib/python3.8/site-packages/numpy-1.16.4-py3.8-linux-x86_64.egg/numpy/__init__.py:145:4304-4320)
Original Internal Error: 
java.lang.NullPointerException
	at com.oracle.graal.python.builtins.objects.cext.capi.CApiContext.getTraceMallocDomain(CApiContext.java:233)
...snip...
```
[ShareEngine.output.r8.txt](https://github.com/graalvm/graalpython/files/5657681/ShareEngine.output.r8.txt)

Thanks!

**Comment by timfel** [comment]
Thank you for the report. We have recently been made aware of this issue on our public Slack channel. A fix is already in master and due to be released with the next release (21.0).

Thanks for trying out these things and please don't hesitate to report any further issues you may find :)

**Comment by kariya-mitsuru** [comment]
Thank you for your quick response and sorry for wasting your time.
I would like to try master branch immediately.

I can't wait for the release of 21.0!
